### DISCUSSION TITLE
System.AggregateException - None of the brokers could be reached (No such host is known.)

### DISCUSSION THREAD

**Discussion by fishberg5051** [rabbitmq/rabbitmq-stream-dotnet-client#214] [Q&A]
Hello,
I'm using RabbitMQ.Client 6.4.0 and RabbitMQ.Stream.Client 1.0.1 to perform a POC.
Following the documentation / examples therein I'm running into an issue I'm having a hard time troubleshooting.
I can create a StreamSystem object, and verify it connects in our rabbit server.  I can verify a stream exists by calling the "system".StreamExists method.

However, when I attempt to create a Consumer, I'm always getting an System.AggregateException - None of the brokers could be reached (No such host is known.).

We are on a windows server, RabbitMQ 3.11.2, Erlang 25.1.
Pulling all our values from a database config.
Our "cluster" only has one node for this experiment.
I have ran the plugin to enable streams, with all the default ports/listeners.

```csharp
var endPointList = new List<EndPoint>();
            foreach (var serverItem in rabbitConfig.RabbitServers)
            {
                endPointList.Add(new IPEndPoint(IPAddress.Parse(serverItem.HostName), 5552));
            }


            //setup the stream config
            var config = new StreamSystemConfig
            {
                UserName = rabbitConfig.UserName,
                Password = rabbitConfig.Password,
                VirtualHost = rabbitConfig.VirtualHost,
                Endpoints = endPointList,
                ClientProvidedName = string.Format("{0}_consumer", rabbitConfig.QueueName)
            };

            var system = await StreamSystem.Create(config); 

            string stream = rabbitConfig.QueueName;

            bool exists = await system.StreamExists(rabbitConfig.QueueName); //returns true

            try
            {
                //TODO: consumer should only be created once per type
                var consumer = await Consumer.Create(
                    new ConsumerConfig(system, stream)
                    {
                        IsSingleActiveConsumer = true,
                        Reference = String.Format("{0}_consumer", stream),
                        // Consume the stream from the beginning 
                        // See also other OffsetSpec 
                        OffsetSpec = new OffsetTypeFirst(), //use the rabbit config to check what type it should use and what value.
                                                            // Receive the messages
                        MessageHandler = async (sourceStream, consumer, ctx, message) =>
                        {
                            try
                            {
                                Console.WriteLine("Got here for consumer");

                            }catch(Exception ex)
                            {
                                _logger.Error(ex.Message);
                            }

                            await Task.CompletedTask;
                        }
                    });;

                var isOpen = consumer.IsOpen();
            }
            catch(Exception ex)
            {
                //Exception is thrown here.
            }
```

I would really welcome any feedback, issues, or corrections on how to solve this.
Please let me know if there are any other details I can provide.

**Comment by Gsantomaggio** [comment]
```
System.AggregateException - None of the brokers could be reached (No such host is known.).
``` 
means that the consumer can't look up the DNS name. 

You need to check if the client can look up server's DNS name.
If there is a proxy or load balancer please read https://github.com/rabbitmq/rabbitmq-stream-dotnet-client#load-balancer

**Reply by fishberg5051** [reply]
@Gsantomaggio  - Thank you kindly for the direction.  I added that line and I was able to consume messages from a stream!
Out of curiosity, I would not consider our setup "Load Balanced".  Lets say in production I have a 3 node cluster....would it matter what IP address is used for the load balance setting?

Thank you again.  

Cheers.

**Comment by Gsantomaggio** [comment]
@fishberg5051 Great that you solved it.

The client should be able to look up the server's name somehow. This [part of the post ](https://blog.rabbitmq.com/posts/2021/07/connecting-to-streams/#with-a-load-balancer)explains how it works.



> Lets say in production I have a 3 node cluster.

Ok, I suppose the clients can't look up the server's name. In this case, you can pick one.  It would be better to have an LB/Proxy with this configuration. 

The current load-balancer configuration accepts only one address atm. We could consider adding more IP/DNS we will think about that.

**Comment by bSirotkovic** [comment]
Hi, we're receiving the same issue, so I'll add here instead of creating a new issue. 
I've read the "Connecting to streams" blog post, and FYI we're using a 3 node cluster without a load balancer.

Since all our node servers are not supposed to be reachable via DNS, there isn't a valid property we can attach in the advertised_host configuration property. Only solution we thought would work is to put the server IP address inside in the advertised_host. The IP address is private of course.

Digging through your code we found out that the underlying issue is the following line in the RoutingClient.cs
`var hostEntry = await Dns.GetHostEntryAsync(broker.Host);`

In the broker.Host we already have a valid (private) IPAddress we would like to use. Can we have an option to bypass the DNS validation (our address will never be DNS valid) and directly use the value set in the advertised_host property in the following CreateClient method.

**Reply by acogoluegnes** [reply]
Why can you use the private IP addresses in the application but not in the `advertised_host` server setting?

If it's really not possible, you can set the `advertised_host` for each node to a pre-defined value, like `node-1`, `node-2`, and `node-3`. Then you can use an address resolver on the client side with a mapping for each `advertised_value` to the appropriate IP address, like `node-1` => IP address 1, etc.

The `advertised_host` values can be anything, as long as client libraries are able to make something out of them, directly or with an abstraction like an address resolver.

**Reply by Gsantomaggio** [reply]
I will fix the client to handle the IP address

**Reply by bSirotkovic** [reply]
The issues that are blocking us are inside the `LookupConnection` method of the `RoutingHelper` class which gets called during creating of a new Producer.

There are two approaches:
a) `clientParameters.AddressResolver == null`
In this case you insist on `Dns.GetHostEntryAsync` which will always fail considering its a private IP address. 

b) `clientParameters.AddressResolver != null`
In this case routing.CreateClient will always be called with the `EndPoint` of our `AddressResolver`. I don't get the point since the idea is not to connect to the `AddressResolver `but to the Node with the stream leader.

In this method our `broker.Host` and `broker.Port` values are valid at the very start (thanks to the use of `advertised_host`) we would just like to be able avoid DNS validation. I see you already have that flag but we cannot override it:

```
// ValidateDns just validate the DNS 
                // it the real world application is always TRUE
                // routing.ValidateDns == false is used just for test
                // it should not change.
                if (routing.ValidateDns)
                {
                    var hostEntry = await Dns.GetHostEntryAsync(broker.Host);
                    endPointNoLb = new IPEndPoint(hostEntry.AddressList.First(), (int)broker.Port);
                }
```

Please advise

**Reply by acogoluegnes** [reply]
> In this method our broker.Host and broker.Port values are valid at the very start (thanks to the use of advertised_host) we would just like to be able avoid DNS validation.

OK, I missed that, sorry.

So, yes, we're working on making the DNS validation work with an IP.

**Reply by bSirotkovic** [reply]
No worries. Thank you!

**Reply by Gsantomaggio** [reply]
@bSirotkovic https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/218

**Comment by Gsantomaggio** [comment]
closed via https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/218
